mytoyota/controller.py

```diff
@@ -72,77 +79,100 @@ async def first_login(self) -> None:
-                    self._uuid = result[CUSTOMERPROFILE].get(UUID)
-                    self._token = token
-                    _LOGGER.debug("Saving token and uuid")
-                    self._token_expiration = datetime.now()
-
-            elif response.status_code == HTTPStatus.BAD_GATEWAY:
-                if retry:
-                    await self._update_token(retry=False)
-                    return
-                raise ToyotaApiError("Servers are overloaded, try again later")
-            else:
-                raise ToyotaLoginError(
-                    f"Login failed, check your credentials! {response.text}"
+            if resp.status_code != HTTPStatus.FOUND:
+                _LOGGER.error(
+                    f"Authorization failed:\n"
+                    f"  Status_Code = {resp.status_code}\n"
+                    f"  Text        = {resp.text}\n"
+                    f"  Headers     = {resp.headers}"
                 )
+                raise ToyotaLoginError("Authorization failed")
+            authentication_code = parse.parse_qs(
+                httpx.URL(resp.headers.get("location")).query.decode()
+            )["code"]
+
+            # Retrieve tokens
+            resp = await client.post(
+                self._access_token_endpoint,
+                headers={"authorization": "basic b25lYXBwOm9uZWFwcA=="},
+                # f"basic {self.BASIC_AUTH_STRING}"},
+                data={
+                    "client_id": "oneapp",
+                    "code": authentication_code,
+                    "redirect_uri": "com.toyota.oneapp:/oauth2Callback",
+                    "grant_type": "authorization_code",
+                    "code_verifier": "plain",
+                },
+            )
+            if resp.status_code != HTTPStatus.OK:
+                _LOGGER.debug(
+                    f"Authorization failed:\n"
+                    f"  Status_Code = {resp.status_code}\n"
+                    f"  Text        = {resp.text}\n"
+                    f"  Headers     = {resp.headers}"
+                )
+                raise ToyotaLoginError("Failed to retrieve required tokens")
+
+            access_tokens: dict[str, Any] = resp.json()
+            assert "access_token" in access_tokens and "id_token" in access_tokens
+
+            self._token = access_tokens["access_token"]
+            self._uuid = jwt.decode(
+                access_tokens["id_token"],
+                algorithms=["RS256"],
+                options={"verify_signature": False},
+                audience="oneappsdkclient",
+            )[
+                "uuid"
+            ]  # Usefully found in toyota_na
+            self._token_expiration = datetime.now() + timedelta(
+                seconds=access_tokens["expires_in"]
+            )
 
-    async def _is_token_valid(self, retry: bool = True) -> bool:
+    def _is_token_valid(self, retry: bool = True) -> bool:
         """Checks if token is valid"""
+        if self._token == None:
```

Comment:
can be simplified with:

`return self._token is not None and self._token_expiration > datetime.now()`